### PR TITLE
feat(nextgen) Auto Pearl

### PR DESCRIPTION
**Pull request by sqlerrorthing**
Aims and throws a pearl at an enemies pearl trajectory
https://imgur.com/a/WwDqBvh (Preview)

### CODE REVIEW

**Comment by sqlerrorthing** [comment]
inspired by vape.gg.

**Comment by sqlerrorthing** [comment]
More preview
https://imgur.com/a/91mMfLm
https://imgur.com/a/QkP3Lab

**Review by 1zun4 — APPROVED** [review]

**Comment by sqlerrorthing** [comment]
[Demonstration](https://imgur.com/a/nPSaRRZ) of “setting the distance to destination” 
Allows you to customize how close a perl should land to another, by hypotential projection

**Comment by sqlerrorthing** [comment]
https://imgur.com/QbE8cd8
^ Pearl targeting with autopearl

**Comment by sqlerrorthing** [comment]
TODO: 
- [x] Sync pearl mode trigger targets with Targets#Combat Targets
- [x] add few modes: Trigger (on any pearl), Target: (a perl sent by a player who is targeted by KillAura or Crystal Aura or Aimbot)

**Comment by sqlerrorthing** [comment]
@1zun4 please see the part with turning and throwing the ender pearl (line 105-134 in ModuleAutoPearl.kt), I think it's too bad as I haven't completely figured out the liquidbounce API to make a better one.

**Comment by 1zun4** [comment]
> @1zun4 please see the part with turning and throwing the ender pearl (line 105-134 in ModuleAutoPearl.kt), I think it's too bad as I haven't completely figured out the liquidbounce API to make a better one.

Sure. I will take a look at it and try it out before I merge it.

**Comment by sqlerrorthing** [comment]
> > @1zun4 please see the part with turning and throwing the ender pearl (line 105-134 in ModuleAutoPearl.kt), I think it's too bad as I haven't completely figured out the liquidbounce API to make a better one.
> 
> Sure. I will take a look at it and try it out before I merge it.

np

**Comment by 1zun4** [comment]
image
Seems like I have no permissions to push to your fork.

File: src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/misc/ModuleAutoPearl.kt
Comment (superblaubeere27) [inline src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/misc/ModuleAutoPearl.kt:239]: What does this function do?

File: src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/misc/ModuleAutoPearl.kt
Comment (superblaubeere27) [inline src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/misc/ModuleAutoPearl.kt:71]: What is this queue for?

**Review by superblaubeere27 — CHANGES_REQUESTED** [review]
This code is good 👍🏻  But there is close to no documentation. Please state what the code sections (especially the sections mentioned below) are supposed to do. Additionally some sections could profit from extracting them in their own functions

File: src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/misc/ModuleAutoPearl.kt
Comment (1zun4) [reply src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/misc/ModuleAutoPearl.kt:239]: My refactor removes this function already - it is useless.

File: src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/misc/ModuleAutoPearl.kt
Comment (1zun4) [reply src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/misc/ModuleAutoPearl.kt:71]: The queue is used to put rotations from PacketEvent to be handled in the Game Tick Handler.

**Comment by 1zun4** [comment]
https://github.com/sqlerrorthing/LiquidBounce/pull/1 Please merge this PR to your repository, after that I can merge this PR.

**Comment by 1zun4** [comment]
Merged via https://github.com/CCBlueX/LiquidBounce/pull/4926

**Comment by sqlerrorthing** [comment]
👍

File: src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/misc/ModuleAutoPearl.kt
Comment (sqlerrorthing) [reply src/main/kotlin/net/ccbluex/liquidbounce/features/module/modules/misc/ModuleAutoPearl.kt:239]: > What does this function do?

yaw, pitch normalization 
-180 <= yaw <= 180, -90 <= pitch <= 90

I didn't understand the code base very well, so I invented the bicycle.

**Pull request closed without merging**
